synapse/storage/persist_events.py

```diff
@@ -735,9 +736,32 @@ def _maybe_start_persisting(self, room_id: str):
                 if event_id == event.event_id:
                     if event.membership == Membership.JOIN:
                         return True
+                else:
+                    events_to_check.append(event_id)
+
+        # Check if any of the changes that we don't have events for are joins.
+        if events_to_check:
+            rows = await self.main_store.get_membership_from_event_ids(events_to_check)
+            is_still_joined = any(row["membership"] == Membership.JOIN for row in rows)
+            if is_still_joined:
+                return True
+
+        # None of the changes to state are joins, so we fall back to checking
+        # the current state of the room to see if any of our users are joined
+        # (ingnoring users that have changed state.)
```

Comment:
"Ingnoring", but I would rephrase for clarity:

```suggestion
        # None of the new state events are joins, so now we check the current
        # room state to see if there are any other users in the room.
```